src/v/cluster/controller.cc

```diff
@@ -797,85 +797,76 @@ ss::future<> controller::shutdown_input() {
-          .then([this] {
-              if (_recovery_backend) {
-                  return _recovery_backend->stop_and_wait();
-              }
-              return ss::make_ready_future();
-          })
-          .then([this] { return _recovery_manager.stop(); })
-          .then([this] { return _recovery_table.stop(); })
-          .then([this] { return _partition_balancer.stop(); })
-          .then([this] { return _metrics_reporter.stop(); })
-          .then([this] { return _feature_manager.stop(); })
-          .then([this] { return _hm_frontend.stop(); })
-          .then([this] { return _hm_backend.stop(); })
-          .then([this] { return _health_manager.stop(); })
-          .then([this] { return _members_backend.stop(); })
-          .then([this] { return _data_migration_worker.stop(); })
-          .then([this] { return _data_migration_frontend.stop(); })
-          .then([this] { return _config_manager.stop(); })
-          .then([this] { return _api.stop(); })
-          .then([this] { return _shard_balancer.stop(); })
-          .then([this] { return _backend.stop(); })
-          .then([this] { return _tp_frontend.stop(); })
-          .then([this] { return _plugin_frontend.stop(); })
-          .then([this] { return _quota_frontend.stop(); })
-          .then([this] { return _ephemeral_credential_frontend.stop(); })
-          .then([this] { return _security_frontend.stop(); })
-          .then([this] { return _members_frontend.stop(); })
-          .then([this] { return _config_frontend.stop(); })
-          .then([this] { return _feature_backend.stop(); })
-          .then([this] { return _bootstrap_backend.stop(); })
-          .then([this] { return _oidc_service.stop(); })
-          .then([this] { return _authorizer.stop(); })
-          .then([this] { return _ephemeral_credentials.stop(); })
-          .then([this] { return _data_migrated_resources.stop(); })
-          .then([this] { return _roles.stop(); })
-          .then([this] { return _credentials.stop(); })
-          .then([this] { return _tp_state.stop(); })
-          .then([this] { return _members_manager.stop(); })
-          .then([this] { return _stm.stop(); })
-          .then([this] { return _quota_backend.stop(); })
-          .then([this] { return _quota_store.stop(); })
-          .then([this] { return _plugin_backend.stop(); })
-          .then([this] { return _plugin_table.stop(); })
-          .then([this] { return _drain_manager.stop(); })
-          .then([this] { return _shard_placement.stop(); })
-          .then([this] { return _partition_balancer_state.stop(); })
-          .then([this] { return _partition_allocator.stop(); })
-          .then([this] { return _partition_leaders.stop(); })
-          .then([this] { return _members_table.stop(); })
-          .then([this] { return _gate.close(); })
-          .then([this] { return _as.stop(); });
+    if (_leader_balancer) {
+        co_await _leader_balancer->stop();
+    }
+
```

Comment:
these kind of things (`c/controller: refactor stop() into a coroutine`) are very good candidates for plucking out into separate pull requests. really as soon as you create the commit you can cherry pick it out and send it up. the upside is then these PRs become much smaller.